### PR TITLE
Improve styles of filtering options for Clippy's lint list

### PR DESCRIPTION
**Pull request by CrazyRoka**
Partially solves #7958 

Updated styles for filtering options. It now uses dropdown menus.

image

changelog: none

### CODE REVIEW

**Comment by camsteffen** [comment]
The dropdowns look really nice! I think there should be some sort of always-visible indicator of what is selected. Possibly just the number of selected items, or maybe even just a different style to indicate that the default selection has been changed.

**Comment by CrazyRoka** [comment]
I added badges to display selected items count. Have a look:

image

**Comment by camsteffen** [comment]
Nice. I think the badge could be the same color as the lint category badges. Also can we see what it looks like on the right side of the button?

**Comment by CrazyRoka** [comment]
I updated the code. Have a look:

- Badge color is now `#777`
- Badge is now located at the right side

image

**Comment by xFrednet** [comment]
Also, a +1 from me for the dropdown menus, they look really nice :+1: 

The website also supports the different color themes from mdBook. The change theme button is in the upper-left corner. Can you make sure that your changes also support the change of color? You can find the available css values here: https://github.com/rust-lang/mdBook/blob/master/src/theme/css/variables.css

**Comment by CrazyRoka** [comment]
@xFrednet @camsteffen Have a look on this video:

https://user-images.githubusercontent.com/19844144/144762393-14fd470f-9ce5-412c-aa03-569c96858d5b.mp4

**Comment by xFrednet** [comment]
I really like it, nice work! :+1: 

Also cc: @flip1995 since you've suggested this tabbed filter layout. :upside_down_face:

**Comment by camsteffen** [comment]
I am seeing JavaScript errors upon selecting options.

The button style seems to change from selected to deselected after clicking one of the options. Maybe this could be fixed with a CSS rule? If not it's probably fine.

**Comment by CrazyRoka** [comment]
> I am seeing JavaScript errors upon selecting options.

I could see the error in Firefox browser, but not in Chrome. Replaced invalid code with correct one.

> The button style seems to change from selected to deselected after clicking one of the options. Maybe this could be fixed with a CSS rule? If not it's probably fine.

There was a style from bootstrap. I replaced it with custom one. Now selected drop-down should be slightly darker than others.

**Comment by camsteffen** [comment]
I think it would be wise to use the Bootstrap dropdown js plugin ([link](https://bootstrapdocs.com/v3.3.6/docs/javascript/#dropdowns)), especially since we're already using Bootstrap CSS. This would be an added dependency though.

@rust-lang/clippy Everyone good with the general direction here?

**Comment by flip1995** [comment]
I like the general approach and how the drop down menus look. I don't know anything about JS / CSS, so I can't really review or give opinions on that front.

Only concern is, if and how it looks on mobile.

File: util/gh-pages/index.html [hunk 1]
```diff
@@ -25,7 +25,63 @@
...
+            color: var(--fg);
+            background: var(--theme-popup-bg);
+            border: 1px solid var(--theme-popup-border);
+        }
+
+        .dropdown-menu .divider {
+            background-color: var(--theme-popup-border);
+        }
+
+        .dropdown-menu .checkbox {
```
Comment (flip1995) [inline util/gh-pages/index.html:38]: There are some trailing whitespaces in this file.

**Review by flip1995 — COMMENTED** [review]
Selecting "All" has a small lag. This was especially noticeable on mobile.

Here's a link to a deployed version of this branch (rebased on current master): https://flip1995.github.io/rust-clippy/master/index.html 

One issue I found on mobile is, that the color scheme drop down is behind the other drop downs. Otherwise it LGTM on mobile as well.
Screenshot_20211206-182635

**Comment by camsteffen** [comment]
The "Filter" field could be on the same line and wrap responsively.

**Comment by CrazyRoka** [comment]
> Selecting "All" has a small lag. This was especially noticeable on mobile.

I can see small lag on my modern laptop. Looks like selecting "All" will render all the lints at once and it causes Angular to take time building the DOM and parsing html:

image

We can add pagination to prevent this

File: util/gh-pages/index.html (hunk 1 above)
Comment (CrazyRoka) [reply util/gh-pages/index.html]: Removed all the trailing whitespaces

**Comment by CrazyRoka** [comment]
> The "Filter" field could be on the same line and wrap responsively.

Implemented this one.

> One issue I found on mobile is, that the color scheme drop down is behind the other drop downs.

Fixed this one

> I think it would be wise to use the Bootstrap dropdown js plugin (link), especially since we're already using Bootstrap CSS. This would be an added dependency though.

Does it support multi-select? As you can see, current dropdown supports selecting multiple entries. Also, it doesn't close dropdown, so user can select multiple items and close it by clicking on the button or somewhere in the page.

**Comment by camsteffen** [comment]
Another thing we could do about the lag is make the lint panel expanded contents render lazily with something like `ngIf="expanded"`. That could be a standalone enhancement.

> Does it support multi-select? As you can see, current dropdown supports selecting multiple entries. Also, it doesn't close dropdown, so user can select multiple items and close it by clicking on the button or somewhere in the page.

I'm only judging by the docs, but I think the dropdown plugin is pretty unopinionated - you can put any HTML in the dropdown menu and you can close it programmatically like you currently are (but I would try out the default behavior first).

**Comment by flip1995** [comment]
Redeployed:  https://flip1995.github.io/rust-clippy/master/index.html

Another thing I noticed on mobile (tangentially related to this PR): The "Fork me on GitHub" banner in the top right corner slightly overlaps with the Lint groups dropdown. So when you try to open the dropdown by pressing on the error it redirects you to the Clippy GH repo, instead of opening the dropdown.

**Comment by CrazyRoka** [comment]
> Another thing I noticed on mobile (tangentially related to this PR): The "Fork me on GitHub" banner in the top right corner slightly overlaps with the Lint groups dropdown. So when you try to open the dropdown by pressing on the error it redirects you to the Clippy GH repo, instead of opening the dropdown.

Fixed this issue with `clip-path` property. So, clickable area would look like this:

image

**Comment by CrazyRoka** [comment]
> Another thing we could do about the lag is make the lint panel expanded contents render lazily with something like `ngIf="expanded"`. That could be a standalone enhancement.

Performance was improved. As you can see, Browser computation is now taking the most time after click:
image

**Comment by CrazyRoka** [comment]
@flip1995 @camsteffen @xFrednet I went through all the reviews, and it looks like all the issues are resolved. Let me know if I missed something.
I haven't used the dropdown plugin, so. The reason is that I implemented this with raw `ng-click`, but this plugin will require me to rewrite it in `jquery` and add additional code to the rendered HTML.
@camsteffen If you strongly suggest using the plugin instead, I will add it.

**Comment by camsteffen** [comment]
Oh I didn't realize it requires jQuery. That's fine then. I hate jQuery lol. I still need to do a full review but I feel confident about this PR overall.

File: util/gh-pages/index.html [hunk 2]
```diff
@@ -240,7 +307,7 @@
 
     </style>
 </head>
-<body>
+<body ng-app="clippy" ng-controller="lintList" ng-click="toggleDropdown(undefined, $event)">
     <div id="theme-icon" class="theme-icon">&#128396;</div>
     <ul id="theme-menu" class="theme-choice" style="display: none;">
```
Comment (xFrednet) [inline util/gh-pages/index.html:312]: The popup menus you created look a bit better than the current theme menu and they also close when the user interacts with the page body besides them. Would you mind updating the theme menu as well as part of this PR?

**Comment by xFrednet** [comment]
The performance increase while searching is wonderful. I really like the changes!

**Review by flip1995 — APPROVED** [review]
This looks really good now. I love the perf improvement. Thank you for addressing all of our reviews so quickly!

Before this can be merged, please rebase your commits, in order to get rid of the merge commit.

File: util/gh-pages/index.html (hunk 2 above)
Comment (CrazyRoka) [reply util/gh-pages/index.html]: I updated theme menu. Now it should close when interacting with the page

**Comment by CrazyRoka** [comment]
> This looks really good now. I love the perf improvement. Thank you for addressing all of our reviews so quickly!
> 
> Before this can be merged, please rebase your commits, in order to get rid of the merge commit.

I rebased my commits and added one new change:
- Theme menu now should be closed when interacting with the page
- Dropdowns are now implemented with directives (less code, more maintainability)
- Removed raw js handler for theme menu

File: util/gh-pages/index.html [hunk 3]
```diff
@@ -160,23 +229,28 @@
...
             padding: 0 0;
+            overflow: hidden;
         }
-        .theme-choice > li {
+
+        .theme-dropdown.open .theme-choice {
+            display: block;
+        }
+
+        .theme-choice>li {
```
Comment (camsteffen) [inline util/gh-pages/index.html:246]: nit: I'd prefer to keep the spacing around `>`

File: util/gh-pages/index.html [hunk 4]
```diff
@@ -25,7 +25,62 @@
...
+        div.panel div.panel-body button.dropdown-toggle:hover {
+            box-shadow: 0 0 3px var(--searchbar-shadow-color);
+        }
+
+        div.panel div.panel-body .open button.dropdown-toggle {
+            background: var(--searchbar-bg);
+            color: var(--searchbar-fg);
+            border-color: var(--theme-popup-border);
+            filter: brightness(90%);
+        }
```
Comment (camsteffen) [inline util/gh-pages/index.html:79]: Bootstrap theming is normally done by downloading `bootstrap-theme.css` and modifying the values. We don't need an entire theme, but I think we should copy parts of that code, and add a comment saying as much.

**Review by camsteffen — COMMENTED** [review]
How about aligning the text and checkboxes like this?

image

Feel free to clean up the commit history a bit as well.

File: util/gh-pages/index.html [hunk 5]
```diff
@@ -25,7 +25,62 @@
...
+         .dropdown-menu .checkbox input {
+             position: relative;
+             margin: 0 0.5rem 0;
+             padding: 0;
+             vertical-align: middle;
+         }
+
+         .dropdown-menu .checkbox:hover {
+            background-color: var(--theme-hover);
+        }
```
[the remark this answers is not in the document]
Comment (CrazyRoka) [reply util/gh-pages/index.html:56]: Fixed it

File: util/gh-pages/index.html (hunk 4 above)
Comment (CrazyRoka) [reply util/gh-pages/index.html:73]: These variables are coming from rust mdBook styles. Which code do you mean?

File: util/gh-pages/index.html [hunk 6]
```diff
@@ -25,7 +25,62 @@
...
+
+        .dropdown-menu .checkbox {
+            width: 100%;
+            display: block;
+            clear: both;
+            font-weight: 400;
+            line-height: 1.42857143;
+            white-space: nowrap;
+            margin: 0;
```
[the remark this answers is not in the document]
Comment (CrazyRoka) [reply util/gh-pages/index.html]: Applied your suggestion

**Comment by CrazyRoka** [comment]
@camsteffen I reviewed your suggestions and applied them. One comment is still under review, have a look. Let me know if you have any other concerns.

**Comment by camsteffen** [comment]
@CrazyRoka just checking in. There's still one open thread. But honestly it's not super important and we can just wrap this up if you prefer.

**Comment by xFrednet** [comment]
Hey @camsteffen, do we want to merge it, since @CrazyRoka hasn't replied yet? I like these changes and would like to have them merged before they conflict with new changes :upside_down_face:

**Comment by camsteffen** [comment]
Okay. @bors r+

**Pull request merged**